libyul/backends/wasm/EVMToEwasmTranslator.cpp

```diff
@@ -806,9 +806,351 @@ function u256_to_address(x1, x2, x3, x4) -> r1, r2, r3 {
+	// slot[0x100..0x200] for state
+
+	let bc_0:i32 := i32.add(slot, 0x1E0:i32) // bc[0]:i64
+	let bc_1:i32 := i32.add(slot, 0x1E8:i32) // bc[1]:i64
+	let bc_2:i32 := i32.add(slot, 0x1F0:i32) // bc[2]:i64
+	let bc_3:i32 := i32.add(slot, 0x1F8:i32) // bc[3]:i64
+	let bc_4:i32 := i32.add(slot, 0x200:i32) // bc[4]:i64
+
+	for { let r:i32 := 0:i32 } i32.lt_u(r, 24:i32) { r := i32.add(r, 1:i32) }
+	{
+		// Theta
+		// for (i = 0; i < 5; i++)
+		//   bc[i] = st[i] ^ st[i + 5] ^ st[i + 10] ^ st[i + 15] ^ st[i + 20];
+		//
+		// bc[0] = st[0x0] ^ st[0x5] ^ st[0xa] ^ st[0xf] ^ st[0x14];
+		i64.store(bc_0, i64.xor(i64.load(i32.add(0x100:i32, slot)),
+						i64.xor(i64.load(i32.add(0x128:i32, slot)),
+						i64.xor(i64.load(i32.add(0x150:i32, slot)),
+						i64.xor(i64.load(i32.add(0x178:i32, slot)),
+								i64.load(i32.add(0x1a0:i32, slot)))))))
+		// bc[1] = st[0x1] ^ st[0x6] ^ st[0xb] ^ st[0x10] ^ st[0x15];
+		i64.store(bc_1, i64.xor(i64.load(i32.add(0x108:i32, slot)),
+						i64.xor(i64.load(i32.add(0x130:i32, slot)),
+						i64.xor(i64.load(i32.add(0x158:i32, slot)),
+						i64.xor(i64.load(i32.add(0x180:i32, slot)),
+								i64.load(i32.add(0x1a8:i32, slot)))))))
+		// bc[2] = st[0x2] ^ st[0x7] ^ st[0xc] ^ st[0x11] ^ st[0x16];
+		i64.store(bc_2, i64.xor(i64.load(i32.add(0x110:i32, slot)),
+						i64.xor(i64.load(i32.add(0x138:i32, slot)),
+						i64.xor(i64.load(i32.add(0x160:i32, slot)),
+						i64.xor(i64.load(i32.add(0x188:i32, slot)),
+								i64.load(i32.add(0x1b0:i32, slot)))))))
+		// bc[3] = st[0x3] ^ st[0x8] ^ st[0xd] ^ st[0x12] ^ st[0x17];
+		i64.store(bc_3, i64.xor(i64.load(i32.add(0x118:i32, slot)),
+						i64.xor(i64.load(i32.add(0x140:i32, slot)),
+						i64.xor(i64.load(i32.add(0x168:i32, slot)),
+						i64.xor(i64.load(i32.add(0x190:i32, slot)),
+								i64.load(i32.add(0x1b8:i32, slot)))))))
+		// bc[4] = st[0x4] ^ st[0x9] ^ st[0xe] ^ st[0x13] ^ st[0x18];
+		i64.store(bc_4, i64.xor(i64.load(i32.add(0x120:i32, slot)),
+						i64.xor(i64.load(i32.add(0x148:i32, slot)),
+						i64.xor(i64.load(i32.add(0x170:i32, slot)),
+						i64.xor(i64.load(i32.add(0x198:i32, slot)),
+								i64.load(i32.add(0x1C0:i32, slot)))))))
+
+		//  for (i = 0; i < 5; i++) {
+		//    t = bc[(i + 4) % 5] ^ ROTL64(bc[(i + 1) % 5], 1);
+		//    for (j = 0; j < 25; j += 5)
+		//      st[j + i] ^= t;
+		//  }
+		// t_0 = bc[4] ^ ROTL64(bc[1], 1)
+		let t_0 := i64.xor(i64.load(bc_4), rotl64(i64.load(bc_1), 1:i64))
+		// t_1 = bc[0] ^ ROTL64(bc[2], 1)
+		let t_1 := i64.xor(i64.load(bc_0), rotl64(i64.load(bc_2), 1:i64))
+		// t_2 = bc[1] ^ ROTL64(bc[3], 1)
+		let t_2 := i64.xor(i64.load(bc_1), rotl64(i64.load(bc_3), 1:i64))
+		// t_3 = bc[2] ^ ROTL64(bc[4], 1)
+		let t_3 := i64.xor(i64.load(bc_2), rotl64(i64.load(bc_4), 1:i64))
+		// t_4 = bc[3] ^ ROTL64(bc[0], 1)
+		let t_4 := i64.xor(i64.load(bc_3), rotl64(i64.load(bc_0), 1:i64))
+
+		// st[0] ^= t_0
+		// st[1] ^= t_1
+		// st[2] ^= t_2
+		// st[3] ^= t_3
+		mstore_internal(i32.add(slot, 0x100:i32),
```

Comment:
Can you write C++ code that generates this code? Looking at this: https://github.com/ethereum/solidity/blob/develop/libsolutil/Keccak256.cpp#L84

Reply:
Or maybe write yul helper functions to return `st[0]`?

Reply:
sure, I will check how this could look like